tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs

```diff
@@ -69,23 +78,86 @@ internal SecurityCoordinator(Security security, HttpContext context, Span span)
     internal IDictionary<string, object> GetPathParams() => _context.Request.RequestContext.RouteData.Values.ToDictionary(c => c.Key, c => c.Value);
 
     /// <summary>
-    /// Initializes a new instance of the <see cref="Coordinator.SecurityCoordinator"/> class.
-    /// framework can do it all at once, but framework only unfortunately
+    /// Framework can do it all at once, but framework only unfortunately
     /// </summary>
     internal void CheckAndBlock(Dictionary<string, object> args)
     {
         var result = RunWaf(args);
         if (result.ShouldBeReported)
         {
+            var reporting = MakeReportingFunction(result.Data, result.AggregatedTotalRuntime, result.AggregatedTotalRuntimeWithBindings);
+
             var blocked = false;
             if (result.ShouldBlock)
             {
-                blocked = WriteAndEndResponse();
-                _httpTransport.MarkBlocked();
+                blocked = ChooseBlockingMethodAndBlock(reporting);
+            }
+
+            reporting(blocked);
+        }
+    }
+
+    private Action<bool> MakeReportingFunction(string triggerData, ulong aggregatedTotalRuntime, ulong aggregatedTotalRuntimeWithBindings)
+    {
+        var securityCoordinator = this;
+        return blocked =>
+        {
+            if (blocked)
+            {
+                securityCoordinator._httpTransport.MarkBlocked();
+            }
+
+            securityCoordinator.Report(triggerData, aggregatedTotalRuntime, aggregatedTotalRuntimeWithBindings, blocked);
+        };
+    }
+
+    private ResponseDetails GetResponseDetails()
+    {
+        string contentType;
+        string body;
+        if (_context.Request.Headers["Accept"] == "text/html")
+        {
+            body = _security.Settings.BlockedHtmlTemplate;
+            contentType = "text/html";
+        }
+        else
+        {
+            body = _security.Settings.BlockedJsonTemplate;
+            contentType = "application/json";
+        }
+
+        return new ResponseDetails() { Body = body, ContentType = contentType };
+    }
+
+    private bool ChooseBlockingMethodAndBlock(Action<bool> reporting)
+    {
+        var isWebApiRequest = _context.CurrentHandler?.GetType()?.FullName == WebApiControllerHandlerTypeFullname;
+        if (isWebApiRequest)
+        {
+            var canThrow = false;
+            lock (_sync)
+            {
+                if (_throwHttpResponseExceptionInitSuccess == null)
+                {
+                    _throwHttpResponseExceptionInitSuccess =
+                        TryCreateThrowHttpResponseExceptionDynMeth(out _throwHttpResponseException);
+                }
+
+                canThrow = _throwHttpResponseExceptionInitSuccess.Value;
```

Comment:
This is a hot path right? Is it worth making this a `Lazy` instead and avoiding the lock?

Reply:
Yes, lazy makes more sense, it simplifies the logic too. Fixed in b9a46c5f6